### PR TITLE
Improved PDBWriter performance by modifying PDBWriter._write_timestep()

### LINKED ISSUE #2785: PDBWriter slow performance
## Expected behavior ##

PDBWriter performs better.

## Actual behavior ##

PDBWriter will take ~20 seconds for a **single** frame with ~300,000 atoms.

## Code to reproduce the behavior ##

``` python
>>> import MDAnalysis as mda
>>> u = mda.Universe(large_pdb)
>>> def write_pdb_file():
>>>     with mda.Writer("test.pdb", n_atoms=u.trajectory.n_atoms) as f:
>>>         f.write(u.atoms)

>>> %lprun -f MDAnalysis.coordinates.PDB.PDBWriter._write_timestep write_pdb_file()
Timer unit: 1e-06 s

Total time: 15.1798 s
Function: _write_timestep at line 978

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   978                                               def _write_timestep(self, ts, multiframe=False):
   979                                                   """Write a new timestep *ts* to file
   980                                           
   981                                                   Does unit conversion if necessary.
   982                                           
   983                                                   By setting *multiframe* = ``True``, MODEL_ ... ENDMDL_ records are
   984                                                   written to represent trajectory frames in a multi-model PDB file. (At
   985                                                   the moment we do *not* write the NUMMDL_ record.)
   986                                           
   987                                                   The *multiframe* = ``False`` keyword signals that the
   988                                                   :class:`PDBWriter` is in single frame mode and no MODEL_
   989                                                   records are written.
   990                                           
   991                                                   .. _MODEL: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#MODEL
   992                                                   .. _ENDMDL: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#ENDMDL
   993                                                   .. _NUMMDL: http://www.wwpdb.org/documentation/file-format-content/format32/sect2.html#NUMMDL
   994                                           
   995                                                   .. versionchanged:: 0.7.6
   996                                                      The *multiframe* keyword was added, which completely determines if
   997                                                      MODEL records are written. (Previously, this was decided based on the
   998                                                      underlying trajectory and only if ``len(traj) > 1`` would MODEL records
   999                                                      have been written.)
  1000                                           
  1001                                                   .. versionchanged:: 1.0.0
  1002                                                      ChainID now comes from the last character of segid, as stated in the documentation.
  1003                                                      An indexing issue meant it previously used the first charater (Issue #2224)
  1004                                           
  1005                                                   """
  1006         1          7.0      7.0      0.0          atoms = self.obj.atoms
  1007         1       5398.0   5398.0      0.0          pos = atoms.positions
  1008         1          4.0      4.0      0.0          if self.convert_units:
  1009         1         13.0     13.0      0.0              pos = self.convert_pos_to_native(pos, inplace=False)
  1010                                           
  1011         1          2.0      2.0      0.0          if multiframe:
  1012         1         22.0     22.0      0.0              self.MODEL(self.frames_written + 1)
  1013                                           
  1014                                                   # Make zero assumptions on what information the AtomGroup has!
  1015                                                   # theoretically we could get passed only indices!
  1016         1          3.0      3.0      0.0          def get_attr(attrname, default):
  1017                                                       """Try and pull info off atoms, else fake it
  1018                                           
  1019                                                       attrname - the field to pull of AtomGroup (plural!)
  1020                                                       default - default value in case attrname not found
  1021                                                       """
  1022                                                       try:
  1023                                                           return getattr(atoms, attrname)
  1024                                                       except AttributeError:
  1025                                                           if self.frames_written == 0:
  1026                                                               warnings.warn("Found no information for attr: '{}'"
  1027                                                                             " Using default value of '{}'"
  1028                                                                             "".format(attrname, default))
  1029                                                           return np.array([default] * len(atoms))
  1030         1       2046.0   2046.0      0.0          altlocs = get_attr('altLocs', ' ')
  1031         1       8645.0   8645.0      0.1          resnames = get_attr('resnames', 'UNK')
  1032         1       2567.0   2567.0      0.0          icodes = get_attr('icodes', ' ')
  1033         1       2439.0   2439.0      0.0          segids = get_attr('segids', ' ')
  1034         1       1261.0   1261.0      0.0          resids = get_attr('resids', 1)
  1035         1        367.0    367.0      0.0          occupancies = get_attr('occupancies', 1.0)
  1036         1        916.0    916.0      0.0          tempfactors = get_attr('tempfactors', 0.0)
  1037         1       5224.0   5224.0      0.0          atomnames = get_attr('names', 'X')
  1038                                           
  1039    299688    3050882.0     10.2     20.1          for i, atom in enumerate(atoms):
  1040    299687     425803.0      1.4      2.8              vals = {}
  1041    299687     679560.0      2.3      4.5              vals['serial'] = util.ltruncate_int(i + 1, 5)  # check for overflow here?
  1042    299687    1521000.0      5.1     10.0              vals['name'] = self._deduce_PDB_atom_name(atomnames[i], resnames[i])
  1043    299687     354330.0      1.2      2.3              vals['altLoc'] = altlocs[i][:1]
  1044    299687     324628.0      1.1      2.1              vals['resName'] = resnames[i][:4]
  1045    299687     326053.0      1.1      2.1              vals['chainID'] = segids[i][-1:]
  1046    299687     877386.0      2.9      5.8              vals['resSeq'] = util.ltruncate_int(resids[i], 4)
  1047    299687     334709.0      1.1      2.2              vals['iCode'] = icodes[i][:1]
  1048    299687     349086.0      1.2      2.3              vals['pos'] = pos[i]  # don't take off atom so conversion works
  1049    299687     319025.0      1.1      2.1              vals['occupancy'] = occupancies[i]
  1050    299687     312598.0      1.0      2.1              vals['tempFactor'] = tempfactors[i]
  1051    299687     369327.0      1.2      2.4              vals['segID'] = segids[i][:4]
  1052    299687    3556776.0     11.9     23.4              vals['element'] = guess_atom_element(atomnames[i].strip())[:2]
  1053                                           
  1054                                                       # .. _ATOM: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#ATOM
  1055    299687    2349730.0      7.8     15.5              self.pdbfile.write(self.fmt['ATOM'].format(**vals))
  1056         1          1.0      1.0      0.0          if multiframe:
  1057         1          3.0      3.0      0.0              self.ENDMDL()
  1058         1          1.0      1.0      0.0          self.frames_written += 1

```

A very quick fix could be changing `for i, atom in enumerate(atoms)` in line 1039 to `for i in range(atoms.n_atoms):` since `atom` is not really used below.

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) '2.0.0-dev0'
- Which version of Python (`python -V`)? 3.8
- Which operating system? Ubuntu 20.04

### PR DESCRIPTION
**Pull request by BFedder**
Fixes #2785

Changes made in this Pull Request:
 - PDBWriter._write_timestep() has improved performance:
   - `enumerate(atoms)` has been replaced with `range(len(atoms))`
   - `_deduce_PDB_atom_names` is now only called on unique pairs of atomnames and resnames, not once for each pair

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

----
📚 Documentation preview 📚: https://mdanalysis--4472.org.readthedocs.build/en/4472/

### CODE REVIEW

File: package/MDAnalysis/coordinates/PDB.py [hunk 1]
```diff
@@ -1217,10 +1217,30 @@ def validate_chainids(chainids, default):
...
+                     for i in range(len(atomnames))]
+
+            unique_pairs = set(pairs)
+
+            deduced_PDB_atom_names = {
+                (atom, res): self._deduce_PDB_atom_name(atom, res)
+                for atom, res in unique_pairs
+                }
+
+            return deduced_PDB_atom_names
```
Comment (tylerjereddy) [inline package/MDAnalysis/coordinates/PDB.py]: Is this not a manual reimplementation of what `@functools.cache` would do for free on `_deduce_PDB_atom_name`? Or maybe I'm missing something

Comment (BFedder) [reply package/MDAnalysis/coordinates/PDB.py]: That's a great suggestion, thanks very much! I wasn't aware of this decorator.

File: package/CHANGELOG [hunk 2]
```diff
@@ -30,6 +30,7 @@ Fixes
  * Fix deploy action to use the correct version of the pypi upload action.
 
 Enhancements
+ * Improved performance of PDBWriter (Issue # 2785, PR # 4472)
```
Comment (RMeli) [inline package/CHANGELOG]:
```suggestion
 * Improved performance of PDBWriter (Issue #2785, PR #4472)
```

**Review by RMeli — APPROVED** [review]

**Comment by RMeli** [comment]
Thanks @BFedder! The failed test was because of a timeout. I fixes the CHANGELOG merge conflicts, so now all tests will re-run.

**Review by yuxuanzhuang — APPROVED** [review]
lgtm!

**Review by IAlibay — APPROVED** [review]

**Comment by IAlibay** [comment]
`mypy` failure is related to something else - CI has passed sucessfully in the past, so I'll go ahead with the merge.

**Pull request merged**